city_scrapers/spiders/il_corrections.py

```diff
@@ -0,0 +1,160 @@
+        print("prase pdf)")
+        lp = LAParams(line_margin=0.1)
+        out_str = StringIO()
+        extract_text_to_fp(BytesIO(response.body), out_str, laparams=lp)
+        pdf_text = out_str.getvalue()
+        return pdf_text.lower()
+
+    def _parse_title(self):
+        """Parse or generate meeting title."""
+        title = "Adult Advisory Board Meeting"
+
+        if "subcommittee" in self.pdf_text:
+            title = "Adult Advisory Board / Women's Subcommittee Meeting"
+        return title
+
+    def _parse_times(self, date, start=True):
+        """Parse start datetime as a naive datetime object."""
+        times = re.findall(r"(\d{1,2}:\d{2} ?(am|a\.m\.|pm|p\.m\.))", self.pdf_text)
+        start_time = times[0][0].replace(".", "")
+        end_time = times[1][0].replace(".", "")
+
+        # Add conversion to datetime object
+        if start:
+            return self._try_time_format(date, start_time)
+        else:
+            return self._try_time_format(date, end_time)
+
+    def _try_time_format(self, date, time):
+        """Try time formatting with and without spacing"""
+        try:
+            time_object = datetime.strptime(f"{date} {time}", "%B %d, %Y %I:%M%p")
+        except ValueError:
+            time_object = datetime.strptime(f"{date} {time}", "%B %d, %Y %I:%M %p")
+
+        return time_object
+
+    def _parse_location(self):
+        """Parse or generate location."""
+
+        location_lookup = {
+            "logan correctional center": {
+                "address": "1096 1350th St, Lincoln, IL 62656",
+                "name": "Logan Correctional Center",
+            },
+            "vandalia correctional center": {
+                "address": "US-51, Vandalia, IL 62471",
+                "name": "Vandalia Correctional Center",
+            },
+            "thompson center": {
+                "address": "100 W. Randolph, Suite 4-200, Chicago, IL 60601",
+                "name": "James R. Thompson Center",
+            },
+            "joliet treatment center": {
+                "address": "2848 McDonough St, Joliet, IL 60431",
+                "name": "Joliet Treatment Center",
+            },
+            "no known location": {"name": "TBD", "address": ""},
+        }
+
+        for location in location_lookup.keys():
+            if location in self.pdf_text:
+                return location_lookup[location]
+
+        return location_lookup["no known location"]
+
+    def _parse_all_links(self, response):
+        """ Gather dates, links """
+        for link in response.css("a"):
+            date = link.re_first(
+                r"""((Jan(uary)?|Feb(ruary)?|Mar(ch)?|Apr(il)?|
+            May|Jun(e)?|Jul(y)?|Aug(ust)?|Sep(tember)?|Oct(ober)?|Nov(ember)?|
+            Dec(ember)?)\s+\d{1,2},\s+\d{4})|((1[0-2]|0?[1-9])/(3[01]|
+            [12][0-9]|0?[1-9])/(?:[0-9]{2})?[0-9]{2})"""
+            )
+
+            if date is not None:
+                if date not in self.links:
```

Comment:
This could be simplified by using `defaultdict` instead of a regular `dict`. Creating this as `defaultdict(dict)` will automatically set the value of any key to a dictionary when it's accessed so you can set `links[date][key]` directly without having to check whether it exists first